csaf-import/src/main/kotlin/io/github/csaf/sbom/retrieval/RetrievedProvider.kt

```diff
@@ -0,0 +1,144 @@
+            ?.mapNotNull { it.directory_url?.toString()?.trimEnd('/') }
+            ?.mapAsync { directoryUrl ->
+                val indexUrl = "$directoryUrl/index.txt"
+                loader
+                    .fetchText(indexUrl)
+                    .fold(
+                        { index ->
+                            index.lines().mapAsync { line ->
+                                val csafUrl = "$directoryUrl/$line"
+                                RetrievedDocument.from(csafUrl, loader, this)
+                            }
+                        },
+                        { e ->
+                            listOf(
+                                Result.failure(
+                                    Exception(
+                                        "Failed to fetch index.txt from directory at $directoryUrl",
+                                        e
+                                    )
+                                )
+                            )
+                        }
+                    )
+            }
+            ?.flatten() ?: emptyList()
+    }
+
+    @Suppress("unused")
+    @JvmOverloads
+    fun fetchDocumentsAsync(loader: CsafLoader = lazyLoader) =
+        ioScope.future { fetchDocuments(loader).map { ResultCompat(it) } }
+
+    companion object {
+        private val ioScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
+
+        @Suppress("unused")
+        @JvmStatic
+        @JvmOverloads
+        fun fromAsync(domain: String, loader: CsafLoader = lazyLoader) =
+            ioScope.future { from(domain, loader).getOrThrow() }
+
+        /**
+         * Retrieves one or more provider-metadata.json documents (represented by the [Provider]
+         * data class) from a domain according to the
+         * [retrieval rules](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#731-finding-provider-metadatajson).
+         */
+        suspend fun from(
+            domain: String,
+            loader: CsafLoader = lazyLoader
+        ): Result<RetrievedProvider> {
+            val ctx = ValidationContext()
+            // Closure for providing HttpResponse to ValidationContext.
+            val ctxEnrichment = { response: HttpResponse -> ctx.httpResponse = response }
+            // TODO: Only the last error will be available in result. We should do some logging.
+            // First, we need to check if a .well-known URL exists.
+            val wellKnownPath = "https://$domain/.well-known/csaf/provider-metadata.json"
+            return loader
+                .fetchProvider(wellKnownPath, ctxEnrichment)
+                .map { it.also { ctx.dataSource = ValidationContext.DataSource.WELL_KNOWN } }
+                .recoverCatching {
+                    // If failure, we fetch CSAF fields from security.txt and try observed URLs
+                    // one-by-one.
+                    loader.fetchSecurityTxtCsafUrls(domain).getOrThrow().firstNotNullOf {
```

Comment:
@milux

Reply:
My suggestion would be that we revisit this in a separate issue. If looking more closely, we might need to rewrite the functionality of discovering the provider metadata, because we are currently discovering and then validating at the only one we found at the end. But actually it seems we need to try, validate and then move to the next step.

Reply:
See https://github.com/csaf-sbom/kotlin-csaf/issues/48